File: scripts/plot_cardiomet_summary.R
args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 2) {
  stop("usage: Rscript plot_cardiomet_summary.R <summary_dir> <plot_dir>")
}

summary_dir <- normalizePath(args[1], mustWork = TRUE)
plot_dir <- args[2]
dir.create(plot_dir, recursive = TRUE, showWarnings = FALSE)

suppressPackageStartupMessages(library(ggplot2))

exposure_order <- c(
  "BMI_GIANT", "BMI_MVP", "WHR_GIANT", "WHRadjBMI_GIANT",
  "SBP_KEATON", "SBP_MVP", "DBP_KEATON", "DBP_MVP", "PP_KEATON",
  "HDL_GLGC", "LDL_GLGC", "TG_GLGC", "TC_GLGC"
)

vascular_outcomes <- c(
  "CHD_FinnGen", "MI_FinnGen", "Stroke_FinnGen",
  "HeartFail_FinnGen", "PADProxy_FinnGen"
)

kidney_outcomes <- c(
  "CKD_CKDGen", "Microalbuminuria_CKDGen", "UACR_CKDGen",
  "RenFail_FinnGen", "KidneyStones_FinnGen", "eGFRCrea_CKDGen", "eGFRCys_CKDGen"
)

run_summary <- read.delim(file.path(summary_dir, "run_summary.tsv"), sep = "\t", check.names = FALSE)
fdr5_hits <- read.delim(file.path(summary_dir, "fdr5_hits.tsv"), sep = "\t", check.names = FALSE)
gene_recurrence <- read.delim(file.path(summary_dir, "gene_recurrence.tsv"), sep = "\t", check.names = FALSE)

heatmap_df <- function(domain_name, outcomes) {
  df <- subset(run_summary, domain == domain_name)
  full <- expand.grid(
    exposure = exposure_order,
    outcome = outcomes,
    stringsAsFactors = FALSE
  )
  merged <- merge(full, df[, c("exposure", "outcome", "fdr5_hits")], by = c("exposure", "outcome"), all.x = TRUE)
  merged$fdr5_hits[is.na(merged$fdr5_hits)] <- 0
  merged$exposure <- factor(merged$exposure, levels = rev(exposure_order))
  merged$outcome <- factor(merged$outcome, levels = outcomes)
  merged
}

plot_heatmap <- function(df, title, out_path) {
  p <- ggplot(df, aes(x = outcome, y = exposure, fill = fdr5_hits)) +
    geom_tile(color = "#f2f2f2", linewidth = 0.4) +
    geom_text(aes(label = fdr5_hits), size = 3.0, colour = "#222222") +
    scale_fill_gradient(low = "#fff7ec", high = "#d7301f") +
    labs(
      title = title,
      x = NULL,
      y = NULL,
      fill = "FDR<5%\nmediators"
    ) +
    theme_minimal(base_size = 11) +
    theme(
      panel.grid = element_blank(),
      axis.text.x = element_text(angle = 35, hjust = 1),
      plot.title = element_text(face = "bold")
    )
  ggsave(out_path, plot = p, width = 9, height = max(4.5, 0.38 * nrow(unique(df["exposure"])) + 1.8), dpi = 220)
}

vascular_df <- heatmap_df("vascular", vascular_outcomes)
kidney_df <- heatmap_df("kidney", kidney_outcomes)

plot_heatmap(
  vascular_df,
  "Vascular outcomes: FDR<5% mediator counts",
  file.path(plot_dir, "vascular_fdr5_heatmap_r.png")
)

plot_heatmap(
  kidney_df,
  "Kidney outcomes: FDR<5% mediator counts",
  file.path(plot_dir, "kidney_fdr5_heatmap_r.png")
)

recurrence_path <- file.path(plot_dir, "top_recurrent_genes_fdr5_r.png")
if (nrow(gene_recurrence) == 0) {
  p <- ggplot() +
    annotate("text", x = 0, y = 0, label = "No FDR<5% mediators detected", size = 6) +
    xlim(-1, 1) +
    ylim(-1, 1) +
    labs(title = "Most recurrent FDR<5% mediator genes") +
    theme_void(base_size = 12) +
    theme(plot.title = element_text(face = "bold", hjust = 0.5))
  ggsave(recurrence_path, plot = p, width = 8, height = 4.5, dpi = 220)
} else {
  top <- head(gene_recurrence, 20)
  top$gene <- factor(top$gene, levels = rev(top$gene))
  p <- ggplot(top, aes(x = fdr5_run_count, y = gene)) +
    geom_col(fill = "#2c7fb8") +
    geom_text(aes(label = fdr5_run_count), hjust = -0.15, size = 3) +
    labs(
      title = "Most recurrent FDR<5% mediator genes",
      x = "Number of exposure-outcome runs",
      y = NULL
    ) +
    theme_minimal(base_size = 11) +
    theme(plot.title = element_text(face = "bold"))
  ggsave(recurrence_path, plot = p, width = 8.5, height = 6, dpi = 220)
}

File: scripts/plot_cardiomet_summary.py
from __future__ import annotations

import argparse
import csv
import os
from collections import Counter
from pathlib import Path

os.environ.setdefault(
    "MPLCONFIGDIR",
    "/Users/duttad5/RFMediation/.codex_tasks/cardiomet_summary/.mplconfig",
)

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt


EXPOSURE_ORDER = [
    "BMI_GIANT",
    "BMI_MVP",
    "WHR_GIANT",
    "WHRadjBMI_GIANT",
    "SBP_KEATON",
    "SBP_MVP",
    "DBP_KEATON",
    "DBP_MVP",
    "PP_KEATON",
    "HDL_GLGC",
    "LDL_GLGC",
    "TG_GLGC",
    "TC_GLGC",
]

VASCULAR_OUTCOMES = [
    "CHD_FinnGen",
    "MI_FinnGen",
    "Stroke_FinnGen",
    "HeartFail_FinnGen",
    "PADProxy_FinnGen",
]

KIDNEY_OUTCOMES = [
    "CKD_CKDGen",
    "Microalbuminuria_CKDGen",
    "UACR_CKDGen",
    "RenFail_FinnGen",
    "KidneyStones_FinnGen",
    "eGFRCrea_CKDGen",
    "eGFRCys_CKDGen",
]


def read_tsv(path: Path) -> list[dict[str, str]]:
    with path.open() as handle:
        return list(csv.DictReader(handle, delimiter="\t"))


def matrix_from_runs(run_rows: list[dict[str, str]], outcomes: list[str]) -> dict[str, list[int]]:
    matrix = {}
    for exposure in EXPOSURE_ORDER:
        vals = []
        for outcome in outcomes:
            row = next((r for r in run_rows if r["exposure"] == exposure and r["outcome"] == outcome), None)
            vals.append(int(row["fdr5_hits"]) if row else 0)
        matrix[exposure] = vals
    return matrix


def plot_heatmap(matrix: dict[str, list[int]], outcomes: list[str], title: str, out_path: Path) -> None:
    exposures = [e for e in EXPOSURE_ORDER if e in matrix]
    data = [matrix[e] for e in exposures]
    vmax = max(max(row) for row in data) if data else 1
    fig, ax = plt.subplots(figsize=(1.35 * len(outcomes) + 3.5, 0.42 * len(exposures) + 2.2), dpi=180)
    im = ax.imshow(data, cmap="YlOrRd", aspect="auto", vmin=0, vmax=max(vmax, 1))
    ax.set_xticks(range(len(outcomes)))
    ax.set_xticklabels(outcomes, rotation=35, ha="right")
    ax.set_yticks(range(len(exposures)))
    ax.set_yticklabels(exposures)
    ax.set_title(title, fontsize=13, weight="bold")
    for i in range(len(exposures)):
        for j in range(len(outcomes)):
            v = data[i][j]
            ax.text(j, i, str(v), ha="center", va="center", fontsize=7, color="#1f1f1f")
    cbar = fig.colorbar(im, ax=ax, fraction=0.03, pad=0.02)
    cbar.set_label("FDR<5% mediators")
    fig.tight_layout()
    fig.savefig(out_path, bbox_inches="tight")
    plt.close(fig)


def plot_top_genes(hit_rows: list[dict[str, str]], out_path: Path) -> None:
    counter = Counter(r["gene"] for r in hit_rows if r.get("gene"))
    top = counter.most_common(20)
    genes = [g for g, _ in top][::-1]
    counts = [c for _, c in top][::-1]
    fig, ax = plt.subplots(figsize=(9.5, 6.5), dpi=180)
    ax.barh(genes, counts, color="#2c7fb8")
    for y, c in enumerate(counts):
        ax.text(c + 0.1, y, str(c), va="center", fontsize=8)
    ax.set_xlabel("Number of exposure-outcome runs with FDR<5% mediation")
    ax.set_title("Most recurrent FDR<5% mediator genes", fontsize=13, weight="bold")
    fig.tight_layout()
    fig.savefig(out_path, bbox_inches="tight")
    plt.close(fig)


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--summary-dir", required=True)
    ap.add_argument("--plot-dir", required=True)
    args = ap.parse_args()

    summary_dir = Path(args.summary_dir)
    plot_dir = Path(args.plot_dir)
    plot_dir.mkdir(parents=True, exist_ok=True)

    run_rows = read_tsv(summary_dir / "run_summary.tsv")
    hit_rows = read_tsv(summary_dir / "fdr5_hits.tsv")

    vascular_runs = [r for r in run_rows if r["domain"] == "vascular"]
    kidney_runs = [r for r in run_rows if r["domain"] == "kidney"]

    plot_heatmap(
        matrix_from_runs(vascular_runs, VASCULAR_OUTCOMES),
        VASCULAR_OUTCOMES,
        "Vascular outcomes: FDR<5% mediator counts",
        plot_dir / "vascular_fdr5_heatmap.png",
    )
    plot_heatmap(
        matrix_from_runs(kidney_runs, KIDNEY_OUTCOMES),
        KIDNEY_OUTCOMES,
        "Kidney outcomes: FDR<5% mediator counts",
        plot_dir / "kidney_fdr5_heatmap.png",
    )
    plot_top_genes(hit_rows, plot_dir / "top_recurrent_genes_fdr5.png")


if __name__ == "__main__":
    main()
